content/blog/comparing-to-kill-a-mockingbird-to-its-sequel-with-apache-spark.md

```diff
@@ -0,0 +1,212 @@
+
+Currently, Spark does not support a user-supplied threshold for Naïve Bayes. Only the best classification rate in the training data is reported. But in real business problems there is an overhead associated with a misclassification so that the *“best” rate may not be the optimal rate*. It is of keen interest to the business to find the point at which maximum value of correct classifications is realized when accounting for incorrect answers. To do this via Spark, we need to use methods that allow for analysis of the threshold.
+
+Given the number of features (a TF-IDF vector of size 10,000) and the nature of the data, Spark’s tree-based ensemble methods are appropriate. Both Random Forest and Gradient Boosted Trees are available.
+
+```scala
+import org.apache.spark.mllib.tree.RandomForest
+import org.apache.spark.mllib.tree.model.RandomForestModel
+import org.apache.spark.mllib.tree.GradientBoostedTrees
+import org.apache.spark.mllib.tree.configuration.BoostingStrategy
+import org.apache.spark.mllib.tree.model.GradientBoostedTreesModel
+
+// RANDOM FOREST REGRESSION
+val categoricalFeaturesInfo = Map[Int, Int]()
+val numClasses = 2
+val featureSubsetStrategy = "auto"
+val impurity = "variance"
+val maxDepth = 10
+val maxBins = 32
+val numTrees = 50
+val modelRF = RandomForest.trainRegressor(train, categoricalFeaturesInfo, numTrees, featureSubsetStrategy, impurity, maxDepth, maxBins)
+
+// GRADIENT BOOSTED TREES REGRESSION
+val boostingStrategy = BoostingStrategy.defaultParams("Regression")
+boostingStrategy.numIterations = 50
+boostingStrategy.treeStrategy.maxDepth = 5
+boostingStrategy.treeStrategy.categoricalFeaturesInfo = Map[Int, Int]()
+val modelGB = GradientBoostedTrees.train(train, boostingStrategy)
+```
+
+The regression model options (estimating vs. classifying) will produce continuous outputs that can be used to find the right threshold. Both of these methods can be configured with tree depth and number of trees– read the Spark <a target='\_blank'  href='http://spark.apache.org/docs/1.3.0/mllib-ensembles.html'>documentation</a> for details, but general rules of thumb are the following:
+
+* Random Forest – trees are built in parallel and overtraining decreases with more trees so setting this number to be large is a great way to leverage a Hadoop environment. The max depth should be larger than GBT.
+* Gradient Boosted Trees – the number of trees is directly related to overtraining and the trees are not built in parallel. This method can produce some extremely high classification rates on the training data, but set the max depth of trees to be smaller than random forest.
+
+The table below shows the commands to calculate the ROC (Receiver Operating Characteristic) for the Random Forest model – the ROC will tell the real story on the model performance.
```

Comment:
Suggest: to calculate the Receiver Operating Characteristic (ROC) for the Random Forest model.